File: Algorithms/Easy/141.Linked-List-Cycle/LinkedListCycle.go
package easy

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(1)
func hasCycle(head *ListNode) bool {
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			return true
		}
	}

	return false
}

File: Algorithms/Easy/141.Linked-List-Cycle/LinkedListCycle.py
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        dummy = ListNode()
        dummy.next = head
        slow, fast = dummy, head

        while slow and fast and fast.next:
            if slow == fast:
                return True
            slow = slow.next
            fast = fast.next.next

        return False